### DISCUSSION TITLE
How to implement custom type which translates to Postgres Numeric?

### DISCUSSION THREAD

**Discussion by onsails** [SeaQL/sea-orm#1978] [Q&A]
I'm working on an app which operates with [ethnum::U256](https://docs.rs/ethnum/latest/ethnum/struct.U256.html) balances and stores them in Postgres as `NUMERIC(78, 0)`.

From what I see in [value.rs](https://github.com/SeaQL/sea-query/blob/f82ab914e6d5ed020ba26ecdc162ba6478b60c78/src/value.rs#L1111) and my experiments with implementing ValueType for `struct SQLU256(ethnum::U256)`, `ValueType::try_from` still expects it to be matched against `sea_query::Value` enum. If I am getting it right – sea_query maps `NUMERIC(78,0)` to Decimal or BigDecimal and has a chance to lose precision.

Is there an API for creating additional Rust types which map directly to db types without relying on `sea_query::Value enum`?

**Comment by tyt2y3** [comment]
> Is there an API for creating additional Rust types

No, there isn't. May be you can cast them as bytes.

**Reply by onsails** [reply]
Thanks for the suggestion! I will experiment with `create cast` to make it cast to a representation compatible with `U256::from_be_bytes` and report here once it's ready.

As an interim solution, managed it to work like this:
```rust
    #[sea_orm(
        column_type = "Decimal(Some((78, 0)))",
        select_as = "text",
        save_as = "uint_256"
    )]
    pub amount: String,
```

Not ideal for application layer but at least db-side IF conds and aggregate functions work well.
